### PR TITLE
Improve duplicate detection accuracy and reduce false positives

### PR DESCRIPTION
**Pull request by danfeder**
## Summary
- 🎯 Raised semantic similarity threshold from 0.3 to 0.5 to reduce low-quality matches
- 📊 Added combined score floor of 0.45 to filter out noise while preserving exact matches
- 🔍 Expanded metadata comparison from 2 fields to all 7 with weighted importance

## Problem
Based on the duplicate detection improvement plan in `/docs/duplicate-detection-improvement-plan.md`:
- System was generating 105 "low" matches vs only 2 "exact" and 2 "high" matches
- Combined score p50 was ~0.293 (too low)
- Only using 2 of 7 available metadata fields for overlap detection
- Simple title matching without normalization

### 1. Enhanced Similarity Algorithms
- **Title Matching**: Added stop word removal, punctuation normalization, and token set ratio
- **Metadata Overlap**: Now compares all 7 fields with weights:
  - Grade levels & Thematic categories: 20% each
  - Activity type & Cultural heritage: 15% each  
  - Season timing, Main ingredients, Cooking methods: 10% each

### 2. Quality Filtering
- Combined score floor of 0.45 (except for exact matches which always pass)
- Limit to top 10 matches per submission
- Semantic threshold raised to 0.5

### 3. Improved UI
- ReviewDashboard now shows exact/high/medium matches by default
- Low-quality matches hidden behind "Show X more low-quality matches" button
- Color-coded match types for quick visual scanning
- Detailed score breakdown for transparency

## Testing
- ✅ 18 unit tests for similarity functions
- ✅ TypeScript type checking passes
- ✅ ESLint validation passes
- ✅ Deployed to production (version 20)

## Expected Impact
- ~75% reduction in low-quality matches displayed
- Better detection of conceptual duplicates through metadata
- Cleaner reviewer experience with less noise
- p50 combined score should increase from ~0.293 to ~0.45+

## Test Plan
- [ ] Submit a test lesson with known duplicates
- [ ] Verify only high-quality matches appear by default
- [ ] Check that low matches can be expanded if needed
- [ ] Confirm exact duplicates are always shown
- [ ] Monitor submission_similarities table for score distribution

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: supabase/functions/detect-duplicates/index.ts [hunk 1]
```diff
@@ -60,47 +60,147 @@ async function generateContentHash(content: string, metadata: any = {}): Promise
...
+  'its',
+  'of',
+  'on',
+  'that',
+  'the',
+  'to',
+  'was',
+  'will',
+  'with',
+  'the',
```
Comment (danfeder) [inline supabase/functions/detect-duplicates/index.ts:89]: **Duplicate stop word in STOP_WORDS set**

Line 89 contains a duplicate entry for 'the' (also present on line 84). While this won't cause functional issues, it's cleaner to remove the duplicate.

```typescript
const STOP_WORDS = new Set([
  'a', 'an', 'and', 'are', 'as', 'at', 'be', 'by', 'for', 'from',
  'has', 'he', 'in', 'is', 'it', 'its', 'of', 'on', 'that',
  'the', // Remove duplicate on line 89
  'to', 'was', 'will', 'with', 'this', 'these', 'those',
  'what', 'when', 'where', 'which', 'who', 'why', 'how',
]);
```

File: supabase/functions/detect-duplicates/index.ts [hunk 2]
```diff
@@ -60,47 +60,147 @@ async function generateContentHash(content: string, metadata: any = {}): Promise
...
 
-  return intersection.size / union.size;
+  // Normalize strings for comparison
+  const normalize = (item: any) => String(item).toLowerCase().trim();
+  const set1 = new Set(arr1.map(normalize));
+  const set2 = new Set(arr2.map(normalize));
+
+  const intersection = new Set([...set1].filter((x) => set2.has(x)));
+  const union = new Set([...set1, ...set2]);
+
```
Comment (danfeder) [inline supabase/functions/detect-duplicates/index.ts:150]: **Potential division by zero in edge case**

While unlikely in practice, if both arrays are somehow empty but pass the initial length checks, `union.size` could be 0. Consider adding a safety check:

```typescript
const intersection = new Set([...set1].filter((x) => set2.has(x)));
const union = new Set([...set1, ...set2]);

return union.size === 0 ? 0 : intersection.size / union.size;
```

This would prevent any potential division by zero scenarios.

File: src/pages/ReviewDashboard.tsx [hunk 3]
```diff
@@ -222,13 +269,65 @@ export function ReviewDashboard() {
...
-    lesson_id: string;
-    lesson: { title: string };
-    combined_score: number;
-  }> => {
-    return [];
+  // Get lesson titles for similarity matches
+  const [lessonTitles, setLessonTitles] = useState<Record<string, string>>({});
+  const [showLowMatches, setShowLowMatches] = useState<Record<string, boolean>>({});
+
+  useEffect(() => {
```
Comment (danfeder) [inline src/pages/ReviewDashboard.tsx:276]: **Potential performance issue with parallel lesson title fetching**

This effect fetches lesson titles for ALL similarities across ALL submissions every time `submissions` changes. This could result in:

1. **Duplicate API calls** - Same lesson IDs fetched multiple times if they appear in multiple submissions
2. **Large queries** - Could fetch hundreds of lesson titles at once
3. **Unnecessary re-fetches** - Triggers on any submission change, even status updates

Consider optimizing with:
```typescript
const [lessonTitlesCache, setLessonTitlesCache] = useState<Record<string, string>>({});

useEffect(() => {
  const fetchMissingTitles = async () => {
    const missingIds = new Set<string>();
    submissions.forEach((sub) => {
      sub.similarities?.forEach((sim) => {
        if (!lessonTitlesCache[sim.lesson_id]) {
          missingIds.add(sim.lesson_id);
        }
      });
    });

    if (missingIds.size > 0) {
      // Batch fetch only missing titles
      const { data: lessons } = await supabase
        .from('lessons')
        .select('lesson_id, title')
        .in('lesson_id', Array.from(missingIds));

      if (lessons) {
        setLessonTitlesCache(prev => ({
          ...prev,
          ...lessons.reduce((acc, lesson) => {
            acc[lesson.lesson_id] = lesson.title;
            return acc;
          }, {} as Record<string, string>)
        }));
      }
    }
  };

  fetchMissingTitles();
}, [submissions, lessonTitlesCache]);
```

**Review by danfeder — COMMENTED** [review]
## 🎯 Excellent Duplicate Detection Improvements!

This PR successfully addresses the duplicate detection issues outlined in the improvement plan. The changes show thoughtful engineering with significant improvements to accuracy and user experience.

### ✅ **Strengths**

**Algorithm Improvements:**
- **Smart threshold tuning**: Raising semantic similarity from 0.3 → 0.5 and adding 0.45 combined score floor will dramatically reduce noise
- **Enhanced title matching**: Stop word removal and token set ratio is much more robust than simple Jaccard
- **Comprehensive metadata comparison**: Expanding from 2 to 7 fields with proper weighting shows excellent attention to detail
- **Proper field mapping**: Correctly maps submission fields to lesson fields (e.g., `gradeLevels` → `grade_levels`)

**UX Improvements:**
- **Smart filtering**: Hiding low-quality matches by default with expandable option is perfect for reviewer workflow  
- **Visual hierarchy**: Color-coded match types (exact/high/medium/low) enable quick scanning
- **Detailed scoring**: Breakdown of title/content/metadata scores provides transparency
- **Performance optimization**: Parallel fetching of profiles and similarities is excellent

**Code Quality:**
- **Comprehensive test coverage**: 18 unit tests covering edge cases and algorithm behavior
- **Type safety**: Proper TypeScript interfaces for `Similarity` type
- **Error handling**: Null checks and graceful fallbacks throughout

### 🔧 **Minor Issues (Non-blocking)**

- **Duplicate STOP_WORDS entry**: Minor cleanup needed (line 89)
- **Performance optimization opportunity**: Consider caching lesson titles to prevent redundant fetches
- **Edge case safety**: Add division-by-zero protection in Jaccard similarity

### 🏗️ **ESYNYC Compliance**

✅ **11 filters maintained** - No filter changes detected  
✅ **TypeScript usage** - Proper interfaces and type safety  
✅ **No console.log** - Using proper error handling  
✅ **Path aliases** - Following established patterns  

### 📊 **Expected Impact Assessment**

Based on the changes, this should deliver on the promised improvements:
- **~75% reduction in low-quality matches** ✅ (0.45 floor vs 0.3 threshold)  
- **Better conceptual duplicate detection** ✅ (7-field metadata comparison)
- **Cleaner reviewer experience** ✅ (smart filtering + visual hierarchy)
- **Higher quality score distribution** ✅ (p50 should move from ~0.293 to ~0.45+)

This is production-ready code that will significantly improve the duplicate detection system. The comprehensive testing and thoughtful algorithm improvements demonstrate excellent engineering practices.

**Recommendation: This looks great! The inline comments address minor issues, but nothing blocking.**
